# INFO 370 PS 1
*Name: Brian Park*

## 1. Base python (17pt)

### 1-1 Lists (3pt)

#### 1. Create a list ‘movies‘ that contains the names of at least six movies you like

In [138]:
movies = ['Avengers: Endgame', 'Pirates of the Caribbean', 'Iron Man', 'About Time', 'Train to Busan', 'Top Gun: Maverick']
print(movies)

['Avengers: Endgame', 'Pirates of the Caribbean', 'Iron Man', 'About Time', 'Train to Busan', 'Top Gun: Maverick']


#### 2. Use slicing to create a list of three first movies in the list

In [139]:
threeM = movies[0:3]
print(threeM)

['Avengers: Endgame', 'Pirates of the Caribbean', 'Iron Man']


#### 3. Use slicing and list concatenation to create a list of the first two and the last two movies

In [140]:
firstTwo = movies[:2]
lastTwo = movies[-2:]
newList = firstTwo + lastTwo
print(newList)

['Avengers: Endgame', 'Pirates of the Caribbean', 'Train to Busan', 'Top Gun: Maverick']


### 1-2 Loops (4pt)

#### 1. Create a list ‘numbers‘ that is the numbers 70 through 79 in the following manner: create an empty list and add numbers to it in a loop (do not use list comprehension or ‘list‘ function here!)

In [141]:
nums = []
for i in range(70, 80):
    nums.append(i)
nums

[70, 71, 72, 73, 74, 75, 76, 77, 78, 79]

#### 2. Use loop to compute sum of all integers from 1 till 100.

In [142]:
sum = 0
for i in range(1, 101):
    sum += i
sum

5050

#### 3. Assign people to seats. Consider names Adam, Ashin, Inukai, Tanaka, and Ikki; and seats 33, 12, 45, 2 and 17. Print seat assignments by assigning each name to the corresponding seat. 

In [143]:
names = ['Adam', 'Ashin', 'Inukai', 'Tanaka', 'Ikki']
seats = [33, 12, 45, 2, 17]

for i in range(5):
    print(names[i] + ": " + str(seats[i]))

Adam: 33
Ashin: 12
Inukai: 45
Tanaka: 2
Ikki: 17


### 1-3 Comprehensions (4pt)

#### 1. Use list comprehension to create a list of squares of numbers 1..10 (i.e. 1, 4, 9, ..., 100)

In [144]:
sqList = [i*i for i in range(1, 11)]
sqList

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

#### 2. Use dict comprehension to create a dict of numbers 1..10 and their squares (i.e. 1:1, 2:4, ..., 10:100).

In [145]:
sqDict = {i: i*i for i in range(1, 11)}
sqDict

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}

### 1-4 Functions and data transformations (6pt)

#### 1. Write a function that takes in time in the numeric form (i.e. not a string) of HHMM (hoursminutes), and returns it in the numeric form of HH.HH (hours + fractions of hours).

In [146]:
def func(time):
    hour = time // 100
    minute = time % 100
    return (hour + (minute / 60))
func(1015), func(345)

(10.25, 3.75)

## 2.  Work with NYC flights data (63pt)

### 2-1 Setup (4pt)

In [147]:
import numpy as np
import pandas as pd

#### 1. Load the data

In [148]:
dfNYC = pd.read_csv("../data/nycflights13.csv.bz2", sep = "\t")

#### 2. Do basic checks:

##### (a). How many variables (columns) is there in the data? Ensure you know the variables in the data. Keep the documentation nearby.

In [149]:
dfNYC.shape

(336776, 19)

There are 19 variables in the dataset.

##### (b). How many rows of data is there?

There are 336,776 rows in the dataset.

##### (c). Print the first few lines of data. Does it look reasonable?

In [150]:
print(dfNYC.head())

   year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute             time_hour  
0      1400     5      15  2013-01-01T10:00:00Z  
1      141

It quite looks reasonable but is sometimes hard to understand by just looking at the dataframe only with variables and without context.

### 2-2 Basic data exploration (19pt)

#### 1. How many flights out of NYC where there in 2013?

In [151]:
ftNYC = dfNYC[(dfNYC.year == 2013) & (dfNYC.origin.isin(["JFK", "LGA", "EWR"]))]
ftNYC.shape[0]

336776

There are 336,776 flights out of NYC in 2013.

#### 2. How many NYC airports are included in this data? Which airports are these?

In [152]:
dfNYC.origin.unique()

array(['EWR', 'LGA', 'JFK'], dtype=object)

The data includes a total of 3 NYC airports: EWR, LGA, and JFK.

#### 3. Into how many airports did the airlines fly from NYC in 2013?

In [153]:
dfNYC[(dfNYC.year == 2013)].dest.nunique()

105

The airlines flew into 105 airports from NYC in 2013.

#### 4. How many flights were there from NYC to Seattle (airport code SEA)?

In [154]:
dfNYC[(dfNYC.dest == "SEA")].shape[0]

3923

There were 3,923 flights from NYC to Seattle.

#### 5. Were there any flights from NYC to Spokane (GEG)?

In [155]:
dfNYC[(dfNYC.dest == "GEG")].shape[0]

0

Based on the computation above, there were no flights from NYC to Spokane.

#### 6. What is the typical delay of the flights in this data?

In [156]:
newNYC = dfNYC.dropna(axis = 0, subset = ["dep_delay"])
round(newNYC["dep_delay"].mean(), 1)

12.6

I calculated the average value of departure delays (dep_delay) when asked about the typical delay.

#### 7. Did you remember to check how good is the delay variable? Are there missings? Are there any implausible or invalid entries? Go and check this if you haven’t done it already

In [157]:
dfNYC.isna().sum()

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

I remember to use ".isna().sum()" operation which allows us to check how many missing values are in each variable. Based on the computation above, there is a lot of missing values in the delay variables.

#### 8. Where was there room for interpretation with these questions (and answers)? How would that affect decisions based on this data?

I tried to understand those questions correctly to find the best answers by mostly looking at variable names and linking them with the questions. However, I've spent much time making decisions about my answers because of some ambiguity in those questions.

### 2-3 Let’s fly to Atlanta! (15pt)

#### 1. How many flights were there from NYC airports to Atlanta in 2013?

In [158]:
toAtl = dfNYC[(dfNYC.dest == "ATL")]
toAtl.shape[0]

17215

There were 17,215 flights from NYC airports to Atlanta in 2013.

#### 2. How many airlines fly from NYC to Atlanta?

In [159]:
toAtl.carrier.nunique()

7

There were 7 airlines flying from NYC to Atlanta.

#### 3. Which are these airlines (just find the 2-letter abbreviations, but you can also check out Bureau of Transportation Statistics https://www.bts.gov/topics/airlines-and-airports/airline-codes)? How many times did each of these go to Atlanta?

In [160]:
toAtl.carrier.unique()

array(['DL', 'MQ', 'FL', 'EV', '9E', 'WN', 'UA'], dtype=object)

In [161]:
toAtl.carrier.value_counts()

DL    10571
FL     2337
MQ     2322
EV     1764
UA      103
9E       59
WN       59
Name: carrier, dtype: int64

Those 7 airlines are 'DL', 'MQ', 'FL', 'EV', '9E', 'WN', and 'UA'. Based on the result above, DL flew 10,571 times, FL flew 2,337 times, MQ flew 2,322 times, EV flew 1,764 times, UA flew 103 times, 9E flew 59 times, and WN flew 59 times to Atlanta.

#### 4.What was the average flight duration to Atlanta? Use arrival and departure time, and compute the flight time yourself! Do not use the airtime variable! Compare your results with direct flights you find in flight schedules (such as expedia or kayak). Is your result similar to the schedules? Comment/explain.

In [162]:
toAtl = dfNYC[dfNYC.dest == "ATL"].copy()
toAtl = toAtl.dropna(axis = 0, subset = ["arr_time", "dep_time"])
toAtl["duration"] = func(toAtl.arr_time) - func(toAtl.dep_time)
toAtl["duration"].mean()

2.1665609751279304

I dropped all rows that have null values either in "dep_time" or "arr_time" to calculate the accurate average value. After that, I used the function made in 1-4 to convert it into the numeric form of HH.HH. As a result, the average flight duration to Atlanta was about 2.17 hours.

I looked up the direct flight schedule from LGA to Atlanta on Kayak. The duration for most of the direct flights took slightly more than 2 hours but never took more than 3 hours which was quite similar to my results.

#### 5. You may see that some of the flight durations are negative. Why? Can you fix this issue?

In [163]:
toAtl[toAtl.duration < 0].head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
1768   2013      1    2    2259.0            2100      119.0     122.0   
21768  2013      1   25    2122.0            2000       82.0      30.0   
25958  2013      1   30    2233.0            2000      153.0     110.0   
26864  2013      1   31    2112.0            2000       72.0       8.0   
26883  2013      1   31    2139.0            1815      204.0       5.0   

       sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
1768             2332      110.0      DL    1247  N362NW    LGA  ATL   
21768            2243      107.0      DL    1147  N639DL    LGA  ATL   
25958            2243      147.0      DL    1147  N618DL    LGA  ATL   
26864            2243       85.0      DL    1147  N697DL    LGA  ATL   
26883            2050      195.0      DL     926  N309DE    EWR  ATL   

       air_time  distance  hour  minute             time_hour   duration  
1768      124.0       762    21       0  2013-01-03T02:00:00Z -21.616667  
21768     109.0       762    20       0  2013-01-26T01:00:00Z -20.866667  
25958     136.0       762    20       0  2013-01-31T01:00:00Z -21.383333  
26864     127.0       762    20       0  2013-02-01T01:00:00Z -21.066667  
26883     126.0       746    18      15  2013-01-31T23:00:00Z -21.566667

This is because those flights depart at night and arrive the next day morning where the value of arr_time becomes smaller than that of dep_time and generates negative flight durations. We can fix this issue by simply adding 2400 to arr_time to avoid the value of dep_time becoming larger than that of arr_time.

#### 6. What percentage of flights to Atlanta were delayed at arrival by more than 15 minutes?

In [164]:
delayArr = toAtl[toAtl.arr_delay > 15].shape[0]
str(round(((delayArr * 100) / (toAtl.shape[0])), 1)) + "%"

'25.2%'

25.2% of flights to Atlanta were delayed at arrival by more than 15 minutes.

### 2-4 What are these planes? (17pt)

#### 1. Load the planes data. What are the variables? How many planes do we have?

In [165]:
dfPlns = pd.read_csv("../data/planes.csv.bz2", sep = "\t")
dfPlns.columns

Index(['tailnum', 'year', 'type', 'manufacturer', 'model', 'engines', 'seats',
       'speed', 'engine'],
      dtype='object')

The data has 'tailnum', 'year', 'type', 'manufacturer', 'model', 'engines', 'seats',
       'speed', 'engine' as variables.

In [166]:
dfPlns.shape[0]

3322

There are 3,322 planes in the data.

#### 2. What would be the merge key, the variable that can connect a flight in the flights data with a plane in the planes data?

'tailnum' is the merge key and the variable being shared in both datasets that have a unique model number of airplanes that can connect two datasets to each other.

#### 3. Merge the two datasets. Do you want to do inner, full, left, right merge in order to answer the next two questions below?

In [167]:
newAtl = pd.merge(toAtl, dfPlns, on='tailnum', how='inner')
newAtl.sample(3)

year_x  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
6075     2013      4   13     753.0             759       -6.0    1030.0   
872      2013      1   29     857.0             900       -3.0    1122.0   
13761    2013      2   16    1656.0            1700       -4.0    1911.0   

       sched_arr_time  arr_delay carrier  ...             time_hour  duration  \
6075             1030        0.0      DL  ...  2013-04-13T11:00:00Z  2.616667   
872              1135      -13.0      DL  ...  2013-01-29T14:00:00Z  2.416667   
13761            1943      -32.0      DL  ...  2013-02-16T22:00:00Z  2.250000   

       year_y                     type      manufacturer     model  engines  \
6075   1998.0  Fixed wing multi engine            BOEING   757-232        2   
872    2008.0  Fixed wing multi engine            BOEING   737-732        2   
13761  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE  A320-212        2   

       seats speed     engine  
6075     178   NaN  Turbo-fan  
872      149   NaN  Turbo-fan  
13761    182   NaN  Turbo-fan  

[3 rows x 28 columns]

For question 4, I used inner join because we only need flights flying to Atlanta that don't have the data about number seats in the plane. Other than these flights, we can exclude them all.

In [168]:
newAtl = pd.merge(toAtl, dfPlns, on='tailnum', how='inner')

For question 5, I also want to use inner join to filter out only flights flying to Atlanta and display their number of seats.

#### 4. How many flights to Atlanta do we have where we don’t have the data about number seats in the plane?

In [169]:
newAtl.seats.isna().sum()

0

There are 0 fligths to Atlanta that don't have the data about number seats.

#### 5. What was the largest plane in terms of seats that flew from NYC to Atlanta? (or what were the largest planes if there were several equally large planes.) Tell us it’s number of seats, manufacturer and model.

In [170]:
merged = newAtl.sort_values(by=['seats'], ascending = False).head(1)
merged[["seats", "manufacturer", "model"]]

seats manufacturer    model
14083    450       BOEING  747-451

The largest plane has 450 seats and is manufactured by BOEING and its model number is 747-451.

### 2-5 Think about all this (8pt)

#### 1. What are your main concerns related to the analysis above? Were questions, answers, methodology your were able to use, and data good enough?

I think that the data was quite good enough to do several basic analyses. However, I found some ambiguity in the data and questions that should need more explanations to be able to find more clarifying analysis. In doing so, my main concern was to misunderstand the questions due to the lack of context that could mislead to the wrong analysis.

#### 2. Can you envision any business-relevant analysis you might do using this data? Hint: it does not have to be freight business, e.g. journalism is business too.

With this data, I think that we can analyze which airlines or planes flown to certain airports have frequently been delayed. This could bring a business insight to compaines or airlines manufacturing the planes and providing their services by improving their flaws to be more prepared for the flights.

#### 3. Do you see any ethical concerns with your analysis? Is it possible that as a result you put someone in a worse situation, even it they do not deserve it?

One ethical concern could be passengers may not know the planes they were on were tracked by us or analysts without their consent. This could possible put someone in trouble if their flights schedule wanted to be hidden from others.

## How much time did you spend?

I almost spend 4 hours to finish this assignment.